Support as-patterns in pattern synonyms with view patterns

expipiplus1 opened the pull request:
I guess a bit of a WIP, Maybe needs neatening and tests

See https://github.com/goldfirere/th-desugar/issues/174

expipiplus1 commented:
View patterns too, a tiny change on top of as-patterns.

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,71 @@ mk_tuple_dpat name_set =
+    pat' <- lift $ removeWilds pat
+    tell [(name, dPatToDExp pat')]
+    return pat'
+  dsViewP _ _ = fail "View patterns are not supported in th-desugar. Use pattern guards instead."
```
This error message is somewhat misleading, as `th-desugar` now does support view patterns in limited circumstances.

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,71 @@ mk_tuple_dpat name_set =
+                   , Quasi, DsMonad
+                   )
+
+class PatM m where
```
This class deserves some Haddocks.

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,71 @@ mk_tuple_dpat name_set =
+                   , Quasi, DsMonad
+                   )
+
+-- | Desugaring a pattern where any as-patterns are returned in reverse order as
+-- further patterns for case analysis on associated names. View patterns are
+-- also returned by pairing their expression and pattern
```
It would be worth stating that this _only_ used when desugaring the pattern on the RHS of a pattern synonym.

RyanGlScott reviewed (commented):
Thanks for the patch! I haven't had a chance to look at the finer details of how this works, but the general idea looks right.

Two possible ideas for making the generated code a bit more compact:

1. This:
   
   ```hs
   pattern P x <- (id -> x)
   ```
   
   Currently desugars to this:
   
   ```hs
   pattern P x_1 <- (\arg_2 -> case arg_2 of
                                   viewed_3 -> case GHC.Base.id viewed_3 of
                                                   x_1 -> x_1 -> x_1)
   ```
   
   The first `case` expression isn't necessary, as it just binds a simple `viewed_3`. Nor is `arg_2` used anywhere in the remainder of the expression. I think it should be possible to check if the very first pattern to be bound (`viewed_3`, in this example) is a bare `DVarP`, and if so, elide the intermediate `case` expression.


2. Regarding your "uncomfortable unpacking/repacking" comment at https://github.com/goldfirere/th-desugar/issues/174#issuecomment-1382674047, what if instead of desugaring this:
   
   ```hs
   data Foo a b = Foo a b
   
   pattern P x a b <- x@(Foo a b)
   ```
   
   To this (cleaned up slightly):
   
   ```hs
   pattern P x_1 a_2 b_3 <- ((\x_1 -> case x_1 of Foo a_2 b_3 -> (x_1, a_2, b_3)) -> (x_1, a_2, b_3))
   ```
   
   We instead desugared to something like this?
   
   ```hs
   pattern P x_1 a_2 b_3 <- ((\x_1 -> (x_1, x_1)) -> (x_1, Foo a_2 b_3))
   ```
   
   That is, if a pattern doesn't contain any view patterns or as-patterns as subpatterns, then match on it in the pattern to the right of the `->` instead of matching on it in the expression to the left of the `->`. This avoids having to unpack it in the expression on the left only to repack it again for the benefit of the pattern on the right.

I haven't thought deeply about how to implement these ideas, so I'm not sure if they work well with the current approach to desugaring patterns.

expipiplus1 commented:
ok, I think I've addressed the comments, as well as added view pattern support everywhere (it fell out very naturally)

I've keps `dsPatX` around for a deprecation cycle if we care about that.

expipiplus1 commented:
err, actually I think there's a bug here I still need to sort!

expipiplus1 commented:
The bug here is that now `dsPatOverExp` return a less fallible pattern, having moved most of the matching on `as-patterns` into the subsequent let expression. The consequence of this is apparent in the following:

```
do 
  let xs = [1,2]
  x@1 <- xs
  pure x
```

This used to desugar to the equivalent of:

```  
do 
  let xs = [1,2]
  1 <- xs
  let x = 1
  pure x
 ```

But now it desugars to the equivalent of:

```
do
  let xs  = [1,2]
  x <- xs
  let 1 = x
  pure x
```

Apart from it being incorrect, I do prefer the second one as there isn't any pattern -> expression transmutation happening.

expipiplus1 commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,71 @@ mk_tuple_dpat name_set =
+-- | Desugar a pattern, returning a list of (Name, DExp) pairs of extra
+-- variables that must be bound within the scope of the pattern
+dsPatC :: DsMonad q => Pat -> q (DPat, [(DExp, DPat)])
+dsPatC = fmap (second reverse) . runWriterT . unPatWithCases . dsPat
```
It was probably cases, I've changed the name to `dsPat'` as it's now the canonical way of desugaring patterns.

RyanGlScott commented:
OK. I could take a look, but there is currently a [build failure](https://github.com/goldfirere/th-desugar/actions/runs/3926783337/jobs/6712823624#step:18:35) in the patch, as caught by CI:

```
[ 9 of 14] Compiling Language.Haskell.TH.Desugar.Core ( Language/Haskell/TH/Desugar/Core.hs, /__w/th-desugar/th-desugar/dist-newstyle/build/x86_64-linux/ghc-9.2.2/th-desugar-1.14/build/Language/Haskell/TH/Desugar/Core.o, /__w/th-desugar/th-desugar/dist-newstyle/build/x86_64-linux/ghc-9.2.2/th-desugar-1.14/build/Language/Haskell/TH/Desugar/Core.dyn_o )

Language/Haskell/TH/Desugar/Core.hs:806:22: error:
    Variable not in scope:
      dsPatWithViewPatterns :: Pat -> q (DPat, t0 (DPat, DExp))
    |
806 |   (pat', subPats) <- dsPatWithViewPatterns pat
    |                      ^^^^^^^^^^^^^^^^^^^^^

Language/Haskell/TH/Desugar/Core.hs:950:19: error:
    Variable not in scope:
      dsPatWithViewPatterns :: Pat -> q (DPat, [(DPat, DExp)])
    |
950 |   (pat', vars) <- dsPatWithViewPatterns pat
    |                   ^^^^^^^^^^^^^^^^^^^^^
```

expipiplus1 commented:
Ah, I didn't mean that I had uncovered an existing bug; I've introduced a new one. Before the "extra patterns" were designed/assumed to be infallible, but now they're not which leads to problems in do notation.

It shouldn't be the end of the world to fix, although it's tempting just to reinstate the old "no view patterns" behaviour.

RyanGlScott commented:
> It shouldn't be the end of the world to fix, although it's tempting just to reinstate the old "no view patterns" behaviour.

OK. For what it's worth, I would be fine with reinstating the old behavior, opening a separate issue to track the remaining tasks.

expipiplus1 commented:
Ok, I reverted to the previous behaviour and opened https://github.com/goldfirere/th-desugar/issues/177

expipiplus1 commented:
I also snuck in a deprecation notice on `dsPatX` hoping to encourage people to use the more versatile `dsPat'`.

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,89 @@ mk_tuple_dpat name_set =
 dsPatOverExp :: DsMonad q => Pat -> DExp -> q (DPat, DExp)
 dsPatOverExp pat exp = do
-  (pat', vars) <- runWriterT $ dsPat pat
+  (pat', vars) <- runPatWithExp $ dsPat pat
```
Is this equivalent to just calling `dsPatX` here?

```suggestion
  (pat', vars) <- dsPatX pat
```

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -876,9 +954,9 @@ dsLetDec (FunD name clauses) = do
   clauses' <- dsClauses (FunRhs name) clauses
   return ([DFunD name clauses'], id)
 dsLetDec (ValD pat body where_decs) = do
-  (pat', vars) <- dsPatX pat
+  (pat', vars) <- dsPat' pat
```
This changes the way that `let` bindings are desugared, which is more ambitious than the title of the PR would suggest. I haven't thought deeply about this change, but could this potentially run into issues similar to the ones documented in #177? Or is this change fine because `let` bindings don't treat fallible patterns the same way that `do`-notation does?

expipiplus1 replied:
Good catch, it does actually change the meaning, however neither are always correct!

I've opened https://github.com/goldfirere/th-desugar/issues/178

expipiplus1 replied:
I'll keep the existing behaviour for now.

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,89 @@ mk_tuple_dpat name_set =
-dsPatX = runWriterT . dsPat
+dsPatX = runPatWithExp . dsPat
+
+-- | Desugar a pattern, returning a list of (DExp, DPat) pairs of extra
```
The comment here says `(DExp, DPat)`, but the type uses `(DPat, DExp)`. I don't have a strong opinion about which order to use, but it would be good to pick a single convention and use it consistently.

expipiplus1 replied:
Ah, the comment is wrong. I settled `DPat` on the left and `DExp` on the right as that's usually the way we see them in the source

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,89 @@ mk_tuple_dpat name_set =
 dsPatsOverExp :: DsMonad q => [Pat] -> DExp -> q ([DPat], DExp)
 dsPatsOverExp pats exp = do
-  (pats', vars) <- runWriterT $ mapM dsPat pats
+  (pats', vars) <- runPatWithExp $ mapM dsPat pats
```
This is _almost_ `dsPatX`, although the types don't quite match up. It might be worth leaving a comment to this effect just in case someone comes along later deciding to address #177, since this is a line of code that would need to be changed in order to accomplish that.

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,89 @@ mk_tuple_dpat name_set =
+runPatWithCases = runWriterT . unPatWithCases
+
+-- | This class permits different implementations of matching on as-patterns
+-- and view-patterns in order to preserve the legacy behavior of 'dsPatX'.
```
Is it really legacy behavior? `dsPatX` is still being used, after all!

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -558,29 +561,89 @@ mk_tuple_dpat name_set =
+    tell [(name, dPatToDExp pat')]
+    return pat'
+  dsViewP _ _ =
+    fail "PatWithExp and dsPatX don't support view-patterns, please use pattern guards instead."
```
I wonder if this error message should mention `PatWithExp` and `dsPatX` at all, given that most users will bump into this restriction without ever directly using either of those things. Perhaps it would be better to say `View patterns aren't supported in patterns in do-notation` instead?

RyanGlScott commented on Language/Haskell/TH/Desugar/Core.hs:
```diff
@@ -747,10 +810,25 @@ dsDec (RoleAnnotD n roles) = return [DRoleAnnotD n roles]
+      -- RHS of the view pattern
+      let p = mkTupleDPat (DVarP <$> allNames)
+      -- Fold up the sub-matches as successive cases
+      let c = foldr (\(asPat, asExp) e -> DCaseE (asExp) [DMatch asPat e])
```
Unnecessary parentheses:

```suggestion
      let c = foldr (\(asPat, asExp) e -> DCaseE asExp [DMatch asPat e])
```

RyanGlScott reviewed (commented):
Thank you so much for your continued work on this!

As one last suggestion, do you feel up to adding some test cases exercising this new functionality?

expipiplus1 commented:
Thank you for taking the time for a thorough review @RyanGlScott!

For the tests, it would be really nice to be able to describe them like so

```haskell
test_t175 :: [Assertion]
test_t175 =
  fmap (uncurry (@=?))
  $(do
    Syn.lift =<< (traverse
      (\(a, e) -> (,) <$> (sweeten @_ @[DDec] <$> (desugar =<< a)) <*> e)
      [ ( [d| pattern P = (id -> ()) |]
        , [d| pattern P = (\x -> case id x of () -> () -> ()) |]
        )
      , ( [d| pattern P x <- x@3 |]
        , [d| pattern P x <- (\x -> case x of 3 -> x -> x) |]
        )
      , ([d| pattern P = ()|], [d| pattern P = ()|])
      ])
   )
```

But with something instead of `@=?` which checks up to alpha-equivalence. I couldn't find anything in th-desugar for this kind of test. Would this be something worth adding or is there a simpler test-setup you'd recommend?

RyanGlScott commented:
Most of the unit tests are located here:

https://github.com/goldfirere/th-desugar/blob/92c07bd88b74efae0d1b6b7bb5d52049723a009a/Test/Run.hs#L77-L154

Where each `test_*` function is located here:

https://github.com/goldfirere/th-desugar/blob/92c07bd88b74efae0d1b6b7bb5d52049723a009a/Test/Splices.hs#L136-L352

The problem with this approach is that it tests things by splicing in _expressions_. However, it's not entirely straightforward to define an expression containing a pattern synonym definition. You can't do something like `let { pattern P = (id -> ()) } in f P = 42`, as pattern synonyms must be defined at the top level.

One way we could accomplish this would be to define the pattern synonym definitions elsewhere, and then define expressions in terms of those pattern synonyms. We would need two copies of each pattern synonym: one using the original pattern synonym syntax, and another using the desugared syntax. We could define them in separate modules (similar to how we have both `Test/Decs.hs` and `Test/DsDecs.hs`) and import them qualified to avoid name clashes.

Testing alpha-equivalence would be another way to accomplish this, but nothing in `th-desugar` quite implements that. The closest that we have is the [`eqTH`](https://github.com/goldfirere/th-desugar/blob/92c07bd88b74efae0d1b6b7bb5d52049723a009a/Test/Splices.hs#L127-L128) function, which checks that two things are `Eq`ual to each other after removing the numeric suffixes from each type variable.

RyanGlScott commented:
How is this going, @expipiplus1? Do you need help with anything?

expipiplus1 commented:
Other life things have gotten in the way a bit! I would like to finalize this, but it's not at the top of my plate at the moment, sorry.

I think there's also a bug, along the lines of: `(foo -> Left x)` being transformed to `(\a -> case foo a of Left x -> x)` (something like that, where fallible matches on the rhs of a view pattern were transformed into infallible (runtime pattern match error) case analysis in the lhs. So very annoyingly we need to add a catch all case to these and handle everything in `Maybe resultTuple` (or some unboxed variant to avoid any allocation).

RyanGlScott commented:
Thanks for the update! (I was worried that this was in a finished state and only being delayed due to figuring out how to add tests, in which case I could pick it up from there.)
